trace_replay/trace_replay.cc

```diff
@@ -764,152 +661,260 @@ Status Replayer::ReadFooter(Trace* footer) {
+    if (!s.ok()) {
+      return s;
+    }
   }
   return TracerHelper::DecodeTrace(encoded_trace, trace);
 }
 
-void Replayer::BGWorkGet(void* arg) {
+Status ReplayerImpl::ToTraceRecord(
+    Trace* trace, std::unordered_map<uint32_t, ColumnFamilyHandle*>* cf_map,
+    int trace_file_version, std::unique_ptr<TraceRecord>* record) {
+  switch (trace->type) {
+    case kTraceWrite:
+      return ToWriteTraceRecord(trace, cf_map, trace_file_version, record);
+    case kTraceGet:
+      return ToGetTraceRecord(trace, cf_map, trace_file_version, record);
+    case kTraceIteratorSeek:
+    case kTraceIteratorSeekForPrev:
+      return ToIterSeekTraceRecord(trace, cf_map, trace_file_version, record);
+    case kTraceMultiGet:
+      return ToMultiGetTraceRecord(trace, cf_map, trace_file_version, record);
+    case kTraceEnd:
+      return Status::Incomplete("Trace end.");
+    default:
+      return Status::NotSupported("Unsupported trace type.");
+  }
+}
+
+Status ReplayerImpl::ExecuteTrace(DB* db,
+                                  std::unique_ptr<TraceRecord>&& record) {
+  switch (record->GetType()) {
+    case kTraceWrite:
+      return ExecuteWriteTrace(db, std::move(record));
+    case kTraceGet:
+      return ExecuteGetTrace(db, std::move(record));
+    case kTraceIteratorSeek:
+    case kTraceIteratorSeekForPrev:
+      return ExecuteIterSeekTrace(db, std::move(record));
+    case kTraceMultiGet:
+      return ExecuteMultiGetTrace(db, std::move(record));
+    case kTraceEnd:
+      return Status::Incomplete("Trace end.");
+    default:
+      return Status::NotSupported("Unsupported trace type.");
+  }
+}
+
+void ReplayerImpl::BackgroundWork(void* arg) {
   std::unique_ptr<ReplayerWorkerArg> ra(
       reinterpret_cast<ReplayerWorkerArg*>(arg));
   assert(ra != nullptr);
-  auto cf_map = static_cast<std::unordered_map<uint32_t, ColumnFamilyHandle*>*>(
-      ra->cf_map);
+
+  std::unique_ptr<TraceRecord> record;
+  Status s = ToTraceRecord(&(ra->trace_entry), ra->cf_map,
+                           ra->trace_file_version, &record);
+  if (s.ok()) {
+    ExecuteTrace(ra->db, std::move(record)).PermitUncheckedError();
+  }
+}
+
+Status ReplayerImpl::ToWriteTraceRecord(
+    Trace* trace, std::unordered_map<uint32_t, ColumnFamilyHandle*>* /*cf_map*/,
+    int trace_file_version, std::unique_ptr<TraceRecord>* record) {
+  assert(trace != nullptr);
+  assert(trace->type == kTraceWrite);
+  if (record == nullptr) {
+    return Status::OK();
+  }
+
+  // To do: is there any validation to convert a payload string to a WriteBatch?
+  if (record != nullptr) {
+    WriteQueryTraceRecord* r = new WriteQueryTraceRecord(trace->ts);
+
+    if (trace_file_version < 2) {
+      r->batch = WriteBatch(trace->payload);
+    } else {
+      WritePayload w_payload;
+      TracerHelper::DecodeWritePayload(trace, &w_payload);
+      r->batch = WriteBatch(w_payload.write_batch_data.ToString());
+    }
+    record->reset(r);
```

Comment:
I think most of this should be done inside of the WriteQueryTraceRecord constructor and not have the fields set externally.  That seems like a cleaner solution and would allow the values to be const.

Reply:
It's intended to leave the values non-const.